File: DataStructures/LinkedLists/leetcode/92. Reverse Linked List II/GPT/ReverseLinkedList_II.js
// ## 1. 問題の分析

// ### 競技プログラミング視点での分析

// * 目標は **left〜right 区間のみ反転**。
// * 追加メモリ無し（O(1)）で **1パス**実現が最速・最小。
// * ダミー節点（sentinel）を用い、`left-1` 番目ノードまで進み、**前挿入(head-insertion)法**で区間を就地反転。

// ### 業務開発視点での分析

// * **境界値**（`left===right` / 先頭からの反転 / 末尾で終わる反転）を網羅。
// * 可読性を保つため、ポインタ名は役割が明確な命名（`pre`, `cur`, `next`）。
// * 例外は**型**と**数値妥当性（left ≤ right, 1 ≤ left）**のみを軽量チェック。

//   * `right ≤ n` の厳密検証は追加パスを要し、1パス要件とトレードオフのため未実施（LeetCode 典型仕様に準拠）。

// ### JavaScript特有の考慮点

// * V8 最適化：単純な `while` / `for` ループと **安定したオブジェクト形状**（ListNode の `{val,next}`）を維持。
// * GC 負荷：新規ノードを**生成しない**（ダミーのみ1個）。
// * 例外スローは**ホットパス前**（冒頭）で実施し、実処理は分岐最小化。

// ---

// ## 2. アルゴリズムアプローチ比較

// | アプローチ                 | 時間計算量 | 空間計算量 | JS実装コスト | 可読性 | 備考               |
// | --------------------- | ----- | ----: | ------: | --- | ---------------- |
// | 方法A: 前挿入で区間反転（1パス・就地） | O(n)  |  O(1) |       低 | 中   | ダミー節点＋3ポインタ。最有力  |
// | 方法B: ノード値を配列に抜き出し反転   | O(n)  |  O(n) |       中 | 高   | 追加メモリが増える（不採用）   |
// | 方法C: スタックで区間反転        | O(n)  |  O(k) |       低 | 中   | k=区間長の追加メモリ（不採用） |

// ---

// ## 3. 選択したアルゴリズムと理由

// * **選択:** 方法A（ダミー節点＋前挿入の就地反転、1パス）
// * **理由:** 速度（O(n)）・メモリ（O(1)）最小で、Follow-up の **one pass** を満たす
// * **JS最適化ポイント:**

//   * ループは `for`/`while` の単純形。
//   * 追加割り当てはダミー節点のみ。
//   * 早期 return と軽量バリデーションでホットパスを汚さない。

// ---

// ## 4. コード実装（solution.js）

// ```javascript
"use strict";
// Module: CommonJS
// Runtime: Node.js v22.14.0（互換: v18+）
// 外部ライブラリ: 不使用（Node標準のみ）

/**
 * Definition for singly-linked list (LeetCode 提供想定)：
 * function ListNode(val, next) { this.val = (val===undefined?0:val); this.next = (next===undefined?null:next); }
 */

/**
 * Reverse Linked List II — 区間 [left, right] を 1-indexed で反転（one pass, in-place）
 * Pure function: 入力リスト構造を就地で反転し、先頭ノード参照を返すのみ（外部副作用なし）
 *
 * @param {ListNode|null} head - 単方向リストの先頭
 * @param {number} left - 反転開始位置 (1-indexed)
 * @param {number} right - 反転終了位置 (1-indexed, left <= right)
 * @returns {ListNode|null} - 反転後の先頭ノード
 * @throws {TypeError} 引数型が不正（head がオブジェクト/ null でない、left/right が有限数でない）
 * @throws {RangeError} left/right の関係 or 下限が不正（left < 1, left > right）
 *
 * 計算量: 時間 O(n), 追加空間 O(1)
 */
function reverseBetween(head, left, right) {
  // --- 入力検証（軽量＆早期） ---
  const isNodeLike = (x) =>
    x === null ||
    (typeof x === "object" && x !== null && "val" in x && "next" in x);
  if (!isNodeLike(head)) throw new TypeError("head must be a ListNode or null");
  if (typeof left !== "number" || !Number.isFinite(left))
    throw new TypeError("left must be a finite number");
  if (typeof right !== "number" || !Number.isFinite(right))
    throw new TypeError("right must be a finite number");
  if ((left | 0) !== left || (right | 0) !== right)
    throw new TypeError("left/right must be integers");
  if (left < 1) throw new RangeError("left must be >= 1");
  if (left > right) throw new RangeError("left must be <= right");

  // 早期終了：空 or 区間長1
  if (head === null || left === right) return head;

  // --- ダミー節点で先頭反転を平滑化 ---
  const dummy = { val: 0, next: head }; // 既存形状 {val,next} に合わせ hidden class を安定化
  let pre = dummy;

  // 1) pre を left-1 まで前進
  for (let i = 1; i < left; i++) {
    // right がリスト長を超えても LeetCode のテストでは想定外だが、防御的に break
    if (pre.next == null) return dummy.next;
    pre = pre.next;
  }

  // 2) 区間内を前挿入で反転
  //    pre.next を区間の先頭（固定）、その直後ノードを先頭直後へ逐次移動
  let cur = pre.next; // 区間の現在先頭（反転後は区間末尾になる）
  let mover = cur ? cur.next : null;

  for (let i = left; i < right && mover != null; i++) {
    // 抜き取り
    cur.next = mover.next;
    // 先頭直後へ差し込み
    mover.next = pre.next;
    pre.next = mover;
    // 次の候補へ
    mover = cur.next;
  }

  return dummy.next;
}

module.exports = { reverseBetween };
// ```

// ---

// ## 5. 追加メモ（JS最適化チェックリスト）

// * ループは `for/while` のみで **コールバック割当無し**（`forEach`/`map` 不使用）。
// * 追加オブジェクトは**ダミー1個のみ**。中間配列・スタック未使用。
// * ListNode 形状（`{val,next}`）を固定し、プロパティ追加や順序変更を行わない。
// * 例外は**ホットパス前**で発生させ、正常系は分岐最小化。
// * 境界パターン（先頭反転・末尾反転・長さ1区間）を一様に処理。

// > **補足（Follow up への回答）**
// > 上記実装はダミー節点＋前挿入により **1パス & O(1) 追加空間** を満たします。

File: DataStructures/LinkedLists/leetcode/92. Reverse Linked List II/GPT/ReverseLinkedList_II.py
# ## 1. 多角的問題分析

# ### 競技プログラミング視点

# * 目標：**一回走査 O(n)**・**追加メモリ O(1)**。
# * 解法：番兵なしでも可能な **「部分区間の原地反転（前向き再配線）」**。`left==1` を個別処理するとダミーノード不要で最小割当。

# ### 業務開発視点

# * 可読性：`pre`（left直前）、`start`（区間先頭）、`curr/prev/next_` を役割明確に命名。
# * 保守性：境界条件 `head is None / left==right / left==1` を早期 return/分岐で処理。
# * 型安全：`Optional["ListNode"]` を徹底。前方参照は文字列リテラルで pylance 互換。

# ### Python特有考慮

# * CPython の関数呼び出し/例外コスト回避のため**ループ内は代入のみ**に限定。
# * 追加ノード（番兵）を作らず**割当 0**にしてメモリフットプリント最小化。
# * 変数はローカルに閉じ、参照寿命を短くし GC 圧力を抑制。

# ---

# ## 2. アルゴリズム比較表

# | アプローチ             | 時間計算量    | 空間計算量    | Python実装コスト | 可読性 | 標準ライブラリ活用 | CPython最適化 | 備考                  |
# | ----------------- | -------- | -------- | ----------- | --- | --------- | ---------- | ------------------- |
# | A: ダミー無し・一回走査（採用） | **O(n)** | **O(1)** | 低           | 高   | 不要        | 良          | `left==1` を個別処理し割当0 |
# | B: ダミー有り・一回走査     | O(n)     | O(1)     | 低           | 高   | 不要        | 良          | 実装容易だが1ノード分の割当      |
# | C: 再帰反転           | O(n)     | O(n)     | 中           | 中   | 不要        | 不可         | 再帰深さ=区間長、不要なコールスタック |

# ---

# ## 3. Python特有最適化ポイント

# * **追加割当をゼロ**：番兵を作らない（`left==1` を専用ロジック）。
# * ループ内は**単純代入のみ**で分岐最小化。
# * 局所変数（`prev/curr/next_`）に束縛して属性アクセス回数を最小化。

# ---

# ## 4. 実装（LeetCode 回答フォーマット / Class 形式）

# ```python
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


from typing import Optional


class Solution:
    """
    Reverse Linked List II
    - One-pass, in-place, no extra node allocation (no dummy).
    - Time: O(n), Space: O(1)
    """

    def reverseBetween(
        self, head: Optional["ListNode"], left: int, right: int
    ) -> Optional["ListNode"]:
        # エッジ：空/区間長1はそのまま
        if head is None or left == right:
            return head

        # Case 1) 先頭から反転（left == 1）: head を直接反転し、新しい先頭を返す
        if left == 1:
            prev: Optional["ListNode"] = None
            curr: Optional["ListNode"] = head
            # 先頭から right 個を反転
            for _ in range(right):
                # 制約 1 <= right <= n より curr は None にならない
                next_: Optional["ListNode"] = curr.next  # type: ignore[reportOptionalMemberAccess]
                curr.next = prev  # type: ignore[reportOptionalMemberAccess]
                prev = curr
                curr = next_
            # 旧先頭 head は反転後の末尾になるので、残り（curr）に接続
            head.next = curr  # type: ignore[reportOptionalMemberAccess]
            return prev

        # Case 2) 中間以降の反転（left > 1）
        # 1) left-1 位置まで進めて pre を「反転開始直前」へ
        pre: "ListNode" = head
        for _ in range(1, left - 1):
            pre = pre.next  # type: ignore[reportOptionalMemberAccess]

        # 2) 区間 [left, right] を通常の反転で反転
        start: "ListNode" = pre.next  # type: ignore[reportOptionalMemberAccess]
        prev: Optional["ListNode"] = None
        curr: Optional["ListNode"] = start
        for _ in range(right - left + 1):
            next_: Optional["ListNode"] = curr.next  # type: ignore[reportOptionalMemberAccess]
            curr.next = prev  # type: ignore[reportOptionalMemberAccess]
            prev = curr
            curr = next_

        # 3) 再接続：pre -> (反転後先頭=prev) ... (反転後末尾=start) -> curr
        pre.next = prev
        start.next = curr
        return head


# ```

# **実装要点**

# * `left==1` と `left>1` を分けることで番兵ノードを廃し、**割当 0** を達成。
# * 反転は「標準の単方向リスト反転」を区間長だけ実施し、その後 **2点で再接続**（`pre.next`, `start.next`）。
# * 型注釈は `Optional["ListNode"]` を使用し、pylance の前方参照問題を回避。`type: ignore` は Optional 成員アクセスの静的誤検知回避に限定。

# ---

# ## 5. 検証（観点のみ）

# * 境界：`[5], (1,1)` / `[1,2], (1,2)` / `[1,2,3], (2,2)` / `[1,2,3,4,5], (1,5)` / `(2,4)` 等で期待形を手計算一致。
# * 制約遵守：一回走査・原地 O(1)・追加割当無し。
# * 型チェック：pylance 前方参照OK、Optionalの扱いはループ構造と制約で安全。
